experimental/WreathMacdonaldpols/src/wreath_macdonald_polynomials.jl

```diff
@@ -0,0 +1,357 @@
+# I would like to credit Dario Mathiä who produced an initial version of the following code
+
+import Chevie: complex_reflection_group, charinfo, CharTable, CyclotomicNumbers
+using Oscar
+import Nemo: IntegerUnion
+
+"""
+Extracted from:
+
+https://github.com/ulthiel/JuLie.jl/blob/master/src/combinatorics/multipartitions.jl
+"""
+# Start of Multipartitions
+
+struct Multipartition{T<:IntegerUnion} <: AbstractArray{Partition{T},1}
+    mp::Array{Partition{T},1}
+end
+
+function multipartition(mp::Vector{Partition{T}}) where T <: IntegerUnion
+  return Multipartition(mp)
+end
+
+function Multipartition(mp::Vector{Vector{T}}) where T <: IntegerUnion
+  return Multipartition([partition(p) for p in mp])
+end
+
+# This is only called when the empty array is part of mp (because then it's
+# "Any" type and not of Integer type).
+function multipartition(mp::Vector{Vector{Any}})
+  return Multipartition([partition(p) for p in mp])
+end
+
+function Base.show(io::IO, ::MIME"text/plain", MP::Multipartition)
+  print(io, MP.mp)
+end
+
+function Base.size(MP::Multipartition)
+  return size(MP.mp)
+end
+
+function Base.length(MP::Multipartition)
+  return length(MP.mp)
+end
+
+function Base.getindex(MP::Multipartition, i::Int)
+  return getindex(MP.mp,i)
+end
+
+function Base.sum(P::Multipartition{T}) where T<:IntegerUnion
+  s = zero(T)
+  for i=1:length(P)
+    s += sum(P[i])
+  end
+  return s
+end
+
+function multipartitions(n::T, r::IntegerUnion) where T<:IntegerUnion
+  MP = Multipartition{T}[]
+
+#We will go through all compositions of n into r parts, and for each such #composition, we collect all the partitions for each of the components of
+  #the composition.
+  #We create the compositions here in place for efficiency.
+
+  #recursively produces all Integer Vectors p of length r such that the sum of all the Elements equals n. Then calls recMultipartitions!
+  function recP!(p::Vector{T}, i::T, n::T) #where T<:IntegerUnion
+    if i==length(p) || n==0
+      p[i] = n
+      recMultipartitions!(fill(Partition(T[]),r), p, T(1))
+    else
+      for j=0:n
+        p[i] = T(j)
+        recP!(copy(p), T(i+1), T(n-j))
+      end
+    end
+  end
+
+  #recursively produces all multipartitions such that the i-th partition sums up to p[i]
+  function recMultipartitions!(mp::Vector{Partition{T}}, p::Vector{T}, i::T) #where T<:IntegerUnion
+    if i == length(p)
+      for q in partitions(p[i])
+        mp[i] = q
+        push!(MP, Multipartition{T}(copy(mp)))
+      end
+    else
+      for q in partitions(p[i])
+        mp[i] = q
+        recMultipartitions!(copy(mp), p, T(i+1))
+      end
+    end
+  end
+
+  recP!(zeros(T,r), T(1), n)
+  return MP
+end
+
+# End of Multipartitions
+
+
+# Tools
+
+#gives indices of a subarray
+function subarray_indices(sub,arr)
+  res = Int[]
+  for i in 1:length(sub)
+    push!(res,findfirst(x->x==sub[i],arr))
+  end
+  return res
+end
+
+# Computes the b-invariant of a partition
+function b1(lambda)
```

Comment:
```suggestion
function b1(lambda::Partition)
```
could you add type signatures like this one to all your methods? That really helps other people understand what these methods are supposed to do, as it documents at least the input types (and it may help to catch some bugs)

Reply:
I have added types in this commit [99b12f3](https://github.com/oscar-system/Oscar.jl/pull/4797/commits/99b12f3bf044c120e2234dd1e32334cddb390ba8).